plan/eliminate_projection.go

```diff
@@ -0,0 +1,128 @@
+// Copyright 2016 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package plan
+
+import (
+	"github.com/pingcap/tidb/expression"
+	"github.com/pingcap/tidb/model"
+)
+
+// EliminateProjection eliminate projection operator to avoid the cost of memory copy in the iterator of projection.
+func EliminateProjection(p LogicalPlan, reorder bool, orderedSchema expression.Schema) LogicalPlan {
+	if reorder {
+		newSchema := make(expression.Schema, len(orderedSchema))
+		for i, v1 := range orderedSchema {
+			for j, v2 := range p.GetSchema() {
+				if v1.TblName == v2.TblName && v1.ColName == v2.ColName {
```

Comment:
'orderedSchema' is constructed according to attribute **Exprs** in projection.
When eliminate a projection whose schema order is different from the child's schema, **only** in this situation, 'orderedSchema' would be reconstructed,  but it would **never** change through the top-to-down process under other circumstances. So fromID/positionID would not be available to compare the equality.